### DISCUSSION TITLE
Save Database As

### DISCUSSION THREAD

**Discussion by flywire** [sqlitebrowser/sqlitebrowser#3676] [Q&A]
Can I open a database, SaveAs another file, Write changes then Save?

**Comment by chrisjlocke** [comment]
Sure .. there is a 'save as' option under the file menu.
Alternatively, as a database is just a file, create a copy of the file
using any file manager.

**Reply by flywire** [reply]
To be clear I have only opened a single sqlite3 db file.

> Sure .. there is a 'save as' option under the file menu.

That is a Save Project As. I wanted to just save the db file with a new name. 

> Alternatively, as a database is just a file, create a copy of the file using any file manager.

Hmm, that means saving original db file to a new file, saving open db to original filename then renaming files.

**Reply by justinclift** [reply]
> Hmm, that means saving original db file to a new file, saving open db to original filename then renaming files.

Are you familiar with copy-and-paste in the windows file manager?

Using that you can create a brand new copy of an existing file.

**Comment by chrisjlocke** [comment]
What version of DB4S are you using?
From at least 3.12.2 the file menu has 'save database as'.

> that means saving original db file to a new file, saving open db to
original filename then renaming files.

Not quite. It means saving db to a new file, opening new file. You're not
saving the open db again as you've just done that. You're not renaming any
files as the save as does that.

Maybe I'm not understanding your problem clearly...

> Message ID:
> ***@***.***
> .com>
>
--
Chris

**Reply by flywire** [reply]
> What version of DB4S are you using? From at least 3.12.2 the file menu has 'save database as'.

Do you mean _Save Project As..._?

![menu](https://github.com/sqlitebrowser/sqlitebrowser/assets/11288701/50fe0d49-fc0d-455c-a649-5055c73c87ba)

DB Browser for SQLite Version 3.12.2

Built for x86_64-little_endian-llp64, running on x86_64
Qt Version 5.12.8
SQLite Version 3.35.5.

> It means saving db to a new file, opening new file.

Which absolutely sucks when you are reviewing a file and find it needs to be changed. _Save As_ is pretty standard functionality in nearly all apps.

**Comment by chrisjlocke** [comment]
Can you try downloading one of the nightly builds. You don't need to
install it .. just download the .zip, extract, and you can run the .exe
directly.
Apologies .. thought 3.12.2 had that feature, but may just be the nightlies.

**Comment by flywire** [comment]
Thanks Chris. It is not what I was expecting.

- Open Database
- change cell
- can't Save Database As
- Write Changes (original database is changed)
- Save Database As

In my view this functionality fails because it does not support saving a database to a new file without changing the original.

- Open Database Read Only
- can't change cell
- Save Database As
- still  can't change cell

Tested with:
```
DB Browser for SQLite Version 3.13.0
Last commit hash when built: b758d90

Built for x86_64-little_endian-llp64, running on x86_64
Qt Version 5.15.2
SQLite Version 3.46.0.
```
and Nightly:
```
DB Browser for SQLite Version 3.13.99 (Jul  3 2024)
Last commit hash when built: 5733cb7

Built for x86_64-little_endian-llp64, running on x86_64
Qt Version 5.15.2
SQLite Version 3.46.0.
```

**Comment by chrisjlocke** [comment]
> It is not what I was expecting

This makes sense when you understand what DB4S is doing. When you change a
cell or perform an update to the database, DB4S runs the necessary SQL on
the current database. SQLite uses transactions, which is where the 'write
changes' button comes in - you're then committing that change.
In Word, which you're comparing, the document you're writing is in memory
and only by clicking Save is that document written to disk from memory.
That's not happening here .. the database isn't in memory.
Technically you could do what you want, but loading a database in memory
could potentially take a long while .. some databases are many GB in size.

Sorry this doesn't work as you want. Unfortunately, does to the way the
product is designed, it's unlikely to change.

The issue with opening a read-only database and 'save as' still being
read-only makes sense, as saving doesn't reload that database (I think).
That functionality could be tweaked easily enough, if others agreed that is
broken.

**Reply by flywire** [reply]
Thank you. See #3677.
